### PR TITLE
Accounts V2: Derived Keymanager, Wallet & Account Creation

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR implements a derived keymanager as part of Prysm's accounts-v2 revamp by conforming to the [EIP-2333](https://eips.ethereum.org/EIPS/eip-2333) and [EIP-2334](https://eips.ethereum.org/EIPS/eip-2334) standards. It implements the following logic:

**Wallet Creation**
1. Asks the user for a wallet password
2. Generates a new random, 32-byte seed for key derivation and prints out a BIP39 mnemonic for the user to acknowledge
3. It stores the encrypted seed to disk under the wallet path with metadata fields such as the next account number

**Account Creation**
1. Asks the user for the wallet password
2. Retrieves and decrypts the seed phrase from disk
3. Asks the user for a new account password
4. Generates a new withdrawal key from the seed at path m/12381/3600/i/0 where i is the account index
5. Generates a new validating key from the seed path m/12381/3600/i/0/0 where i is the account index
6. Logs the newly created account info including public keys to stdout

**Which issues(s) does this PR fix?**

Part of #6220

### CODE REVIEW

File: validator/accounts/v2/wallet.go [hunk 1]
```diff
@@ -28,12 +29,19 @@ const (
 	// WalletDefaultDirName for accounts-v2.
 	WalletDefaultDirName = ".prysm-wallet-v2"
 	// PasswordsDefaultDirName where account passwords are stored.
-	PasswordsDefaultDirName  = ".prysm-wallet-v2-passwords"
-	keymanagerConfigFileName = "keymanageropts.json"
-	passwordFileSuffix       = ".pass"
-	numAccountWords          = 3 // Number of words in account human-readable names.
-	accountFilePermissions   = os.O_CREATE | os.O_RDWR
-	directoryPermissions     = os.ModePerm
+	PasswordsDefaultDirName = ".prysm-wallet-v2-passwords"
```
Comment (rauljordan) [inline validator/accounts/v2/wallet.go:32]: Made these exported for godocs visibility

File: validator/keymanager/v2/derived/derived.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
+package derived
```
Comment (rauljordan) [inline validator/keymanager/v2/derived/derived.go:1]: Review this in depth

File: validator/keymanager/v2/derived/mnemonic.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
+package derived
```
Comment (rauljordan) [inline validator/keymanager/v2/derived/mnemonic.go:1]: Only a derived keymanager needs mnemonic functionality. Will remove the functionality from the direct keymanager in another PR.

File: shared/testutil/log.go [hunk 4]
```diff
@@ -31,6 +31,15 @@ func assertLogs(t *testing.T, hook *test.Hook, want string, flag bool) {
 		if strings.Contains(msg, want) {
 			match = true
 		}
+		for _, field := range e.Data {
```
Comment (rauljordan) [inline shared/testutil/log.go:34]: Check for data in fields as well

File: validator/accounts/v2/wallet.go [hunk 5]
```diff
@@ -269,6 +325,29 @@ func (w *Wallet) ReadPasswordForAccount(accountName string) (string, error) {
 	return string(password), nil
 }
 
+// ReadFileForAccount from the wallet's accounts directory.
+func (w *Wallet) ReadFileForAccount(accountName string, fileName string) ([]byte, error) {
```
Comment (rauljordan) [inline validator/accounts/v2/wallet.go:329]: All of these can likely be abstracted to reduce code duplication, but want to save that for a next PR

File: validator/accounts/v2/wallet_create.go [hunk 6]
```diff
@@ -51,7 +52,13 @@ func CreateWallet(cliCtx *cli.Context) error {
 				"Make a new validator account with ./prysm.sh validator accounts-2 new",
 		)
 	case v2keymanager.Derived:
-		log.Fatal("Derived keymanager is not yet supported")
+		if err = initializeDerivedWallet(cliCtx, walletDir); err != nil {
```
Comment (0xKiwi) [inline validator/accounts/v2/wallet_create.go:55]: This is for creation? `initialize` seems odd here.

File: validator/keymanager/v2/derived/derived.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+	// ValidatingKeyDerivationPathTemplate defining the hierarchical path for validating
+	// keys for Prysm eth2 validators. According to EIP-2334, the format is as follows:
+	// m / purpose / coin_type / account_index / withdrawal_key / validating_key
+	ValidatingKeyDerivationPathTemplate = "m/12381/3600/%d/0/0"
+)
+
+// Wallet defines a struct which has capabilities and knowledge of how
+// to read and write important accounts-related files to the filesystem.
+// Useful for keymanager to have persistent capabilities for accounts on-disk.
+type Wallet interface {
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:47]: Isn't this defined somewhere else already?

File: validator/keymanager/v2/derived/derived.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+	walletSeed := make([]byte, 32)
+	n, err := rand.NewGenerator().Read(walletSeed)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not initialize wallet seed")
+	}
+	if n != len(walletSeed) {
+		return nil, errors.New("could not randomly create seed")
+	}
+	m := &EnglishMnemonicGenerator{
+		skipMnemonicConfirm: false,
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:176]: This is hardcoded, is the command non-interactive?

File: validator/keymanager/v2/derived/derived.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	encryptor := keystorev4.New()
+	cryptoFields, err := encryptor.Encrypt(walletSeed, []byte(password))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encrypt seed phrase into keystore")
+	}
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+	seedFile := &SeedConfig{}
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:194]: Why not define it as a struct like `{Crypto: }`?

File: validator/keymanager/v2/derived/derived.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+	encryptor := keystorev4.New()
+	cryptoFields, err := encryptor.Encrypt(privateKey, []byte(password))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encrypt validating key into keystore")
+	}
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not generate new, random UUID for keystore")
+	}
+	keystoreFile := &Keystore{}
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:300]: same here with the struct

File: validator/accounts/v2/new.go [hunk 11]
```diff
@@ -26,8 +26,8 @@ const (
 )
 
 var keymanagerKindSelections = map[v2keymanager.Kind]string{
-	v2keymanager.Direct:  "Direct, On-Disk Wallet (Recommended)",
-	v2keymanager.Derived: "Derived HD Wallet (Advanced)",
+	v2keymanager.Derived: "Derived, HD Wallet (Recommended)",
+	v2keymanager.Direct:  "Non-HD Wallet (Most Basic)",
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply validator/accounts/v2/new.go:30]: Agreed

File: validator/keymanager/v2/derived/derived.go (hunk 7 above)
Comment (rauljordan) [reply validator/keymanager/v2/derived/derived.go]: Made it a common iface

File: validator/accounts/v2/iface/wallet.go [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"context"
+	"io"
+)
+
+// Wallet defines a struct which has capabilities and knowledge of how
+// to read and write important accounts-related files to the filesystem.
+// Useful for keymanagers to have persistent capabilities for accounts on-disk.
+type Wallet interface {
```
Comment (0xKiwi) [inline validator/accounts/v2/iface/wallet.go:11]: Would be nice to separate these with some comments as // Read functions or // Write functions

File: validator/keymanager/v2/derived/derived.go [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+	"path"
+	"strconv"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	util "github.com/wealdtech/go-eth2-util"
+	keystorev4 "github.com/wealdtech/go-eth2-wallet-encryptor-keystorev4"
+
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:18]:
```suggestion
```

File: validator/keymanager/v2/derived/derived.go [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+	cfg               *Config
+	mnemonicGenerator SeedPhraseFactory
+	keysCache         map[[48]byte]bls.SecretKey
+	lock              sync.RWMutex
+	seedCfg           *SeedConfig
+	seed              []byte
+}
+
+// Keystore json file representation as a Go struct.
+type Keystore struct {
```
Comment (0xKiwi) [inline validator/keymanager/v2/derived/derived.go:64]: Does this need to be defined as well? I think I've seen it elsewhere.

File: validator/accounts/v2/iface/wallet.go [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+	// Methods to retrieve wallet and accounts metadata.
+	AccountNames() ([]string, error)
+	AccountsDir() string
+	CanUnlockAccounts() bool
+	// Read methods for important wallet and accounts-related files.
+	ReadEncryptedSeedFromDisk(ctx context.Context) (io.ReadCloser, error)
+	ReadPasswordForAccount(accountName string) (string, error)
+	ReadFileForAccount(accountName string, fileName string) ([]byte, error)
+	// Write methods to persist important wallt and accounts-related files to disk.
```
Comment (0xKiwi) [inline validator/accounts/v2/iface/wallet.go:20]:
```suggestion
	// Write methods to persist important wallet and accounts-related files to disk.
```

File: validator/accounts/v2/wallet_test.go [hunk 16]
```diff
@@ -10,21 +10,19 @@ import (
 	"path"
 	"testing"
 
+	"github.com/sirupsen/logrus"
+
```
Comment (0xKiwi) [inline validator/accounts/v2/wallet_test.go:14]:
```suggestion
```

File: validator/keymanager/v2/direct/direct.go [hunk 17]
```diff
@@ -307,7 +287,7 @@ func (dr *Keymanager) generateKeystoreFile(validatingKey bls.SecretKey, password
 	if err != nil {
 		return nil, err
 	}
-	keystoreFile := &Keystore{}
+	keystoreFile := &v2keymanager.Keystore{}
```
Comment (0xKiwi) [inline validator/keymanager/v2/direct/direct.go:290]: Can you change this to fill the struct?

File: beacon-chain/state/stateutil/benchmark_test.go [hunk 18]
```diff
@@ -45,17 +45,11 @@ func BenchmarkBlockHTR(b *testing.B) {
 	})
 }
 
-func BenchmarkMerkleize(b *testing.B) {
+func BenchmarkMerkleize_Buffered(b *testing.B) {
```
Comment (rauljordan) [inline beacon-chain/state/stateutil/benchmark_test.go:47]: Updating deps with go mod led to some breaking changes in proto's zszz that are no longer compatible with our old code, so had to amend this function

**Pull request merged**